crates/reporting/src/error/type.rs

```diff
@@ -3126,14 +3155,220 @@ fn diff_record<'b>(
     }
 }
 
+/// This is a helper for should_show_field_diff - see its doc comment for details.
+fn should_show_diff(t1: &ErrorType, t2: &ErrorType) -> bool {
+    use ErrorType::*;
+
+    match (t1, t2) {
+        (Type(sym1, types1), Type(sym2, types2)) => {
+            if sym1 != sym2 || types1.len() != types2.len() {
+                return true;
+            }
+
+            types1
+                .iter()
+                .zip(types2.iter())
+                .any(|(t1, t2)| should_show_diff(t1, t2))
+        }
+        (Infinite, Infinite) | (Error, Error) => false,
+        (FlexVar(v1), FlexVar(v2)) => v1 != v2,
+        (RigidVar(v1), RigidVar(v2)) => v1 != v2,
+        (FlexAbleVar(v1, set1), FlexAbleVar(v2, set2)) => {
+            if v1 != v2 || set1.len() != set2.len() {
+                return true;
+            }
```

Comment:
🤔 If v1==v2, then set1 should always == set2 - maybe worth adding a debug_assert for that instead, and just checking `v1 != v2`